### PR TITLE
fix: permission issue in member tab in workspace 

### PR DESCRIPTION
**Pull request by keita-determined**
## Description
[WEB-973]

### the reported issue test plan
- Create a non-admin user as `workspaceAdmin` to a workspace
- As the non-admin user, the user should be able to change the permission from `workspaceAdmin` to `Editor` in the workspace. This action should not cause any issues, such as losing access to the workspace

### Additional fix
- User can sort by name
- User can search by name and see the correct result
- No `duplicate key` warnings are displayed in console even if there are the same user with different permission

## Checklist

- [x] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[WEB-973]: https://determinedai.atlassian.net/browse/WEB-973?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/utils/user.ts [hunk 1]
```diff
@@ -10,60 +9,30 @@ export function getDisplayName(user: DetailedUser | User | UserNameFields | unde
...
     return user.id;
   }
-  const group = userOrGroup as V1Group;
+  const group = userOrGroup;
 
   // The groupId should always exist
   return group.groupId || 0;
 };
-
-export const getAssignedRole = (
```
Comment (keita-determined) [inline webui/react/src/utils/user.ts:35]: This didnt work if there are the same users with different permission

File: webui/react/src/pages/WorkspaceDetails/WorkspaceMembers.tsx [hunk 2]
```diff
@@ -116,7 +123,40 @@ const WorkspaceMembers: React.FC<Props> = ({
...
+      .flatMap((data) => data?.[1] ?? [])
+      .map((d) => {
+        const userInfo = usersAssignedDirectly.find((u) => u.id === d.userId);
+        const groupWithRole: UserWithRoleInfo = {
+          displayName: userInfo?.displayName,
+          roleAssignment: d.roleAssignment,
+          userId: userInfo?.id ?? -1,
+          username: userInfo?.username ?? '',
+        };
+        return groupWithRole;
```
Comment (julian-determined-ai) [inline webui/react/src/pages/WorkspaceDetails/WorkspaceMembers.tsx:155]: nit: rename this to `userWithRole` or similar.

File: webui/react/src/pages/WorkspaceDetails/WorkspaceMembers.tsx [hunk 3]
```diff
@@ -116,7 +123,40 @@ const WorkspaceMembers: React.FC<Props> = ({
...
+      V1RoleWithAssignments['userRoleAssignments'],
+    ][] = assignments.map((assignment: V1RoleWithAssignments) => {
+      return [assignment.groupRoleAssignments, assignment.userRoleAssignments];
+    });
+    const groups: GroupWithRoleInfo[] = groupsAndUsers
+      .flatMap((data) => data?.[0] ?? [])
+      .map((d) => {
+        const groupnfo = groupsAssignedDirectly.find((g) => g.groupId === d.groupId);
+        const groupWithRole: GroupWithRoleInfo = {
+          groupId: groupnfo?.groupId,
```
Comment (julian-determined-ai) [inline webui/react/src/pages/WorkspaceDetails/WorkspaceMembers.tsx:138]: nit: `groupInfo` instead of `groupnfo`

**Comment by julian-determined-ai** [comment]
@keita-determined could you also update this item in the test place

> As the non-admin user, change the user should be able to change the permission in the workspace workspace to Editor

it's a little hard to understand.

**Comment by keita-determined** [comment]
> @keita-determined could you also update this item in the test place
> 
> > As the non-admin user, change the user should be able to change the permission in the workspace workspace to Editor
> 
> it's a little hard to understand.

sorry, the sentence is totally broken. I changed it to
`As the non-admin user, the user should be able to change the permission from workspaceAdmin to Editor in the workspace. This action should not cause any issues, such as losing access to the workspace`

**Comment by julian-determined-ai** [comment]
@keita-determined One visual bug, if a role is changed and a user switches to another tab and then back, we will briefly show the old role. 

https://user-images.githubusercontent.com/103522725/225349291-81e54f31-c57b-42c1-afd2-cbf927a89e21.mov

**Comment by keita-determined** [comment]
> @keita-determined One visual bug, if a role is changed and a user switches to another tab and then back, we will briefly show the old role.

fixed

**Review by julian-determined-ai — APPROVED** [review]

**Pull request merged**